keps/sig-scheduling/5194-reserved-for-workloads/README.md

```diff
@@ -0,0 +1,796 @@
+type AllocationResult struct {
+  ...
+
+  // ReservedForAnyPod specifies whether the ResourceClaim can be used by
+  // any pod referencing it. If set to true, the kubelet will not check whether
+  // the pod is listed in the staus.ReservedFor list before running the pod.
+  //
+  // +featureGate=DRAReservedForWorkloads
+  // +optional
+  ReservedForAnyPod *bool
+}
+```
+
+The `ResourceClaimConsumerReference type already exists:
+
+```go
+// ResourceClaimConsumerReference contains enough information to let you
+// locate the consumer of a ResourceClaim. The user must be a resource in the same
+// namespace as the ResourceClaim.
+type ResourceClaimConsumerReference struct {
+  // APIGroup is the group for the resource being referenced. It is
+  // empty for the core API. This matches the group in the APIVersion
+  // that is used when creating the resources.
+  // +optional
+  APIGroup string
+  // Resource is the type of resource being referenced, for example "pods".
+  // +required
+  Resource string
+  // Name is the name of resource being referenced.
+  // +required
+  Name string
+  // UID identifies exactly one incarnation of the resource.
+  // +required
+  UID types.UID
+}
+```
+
+#### Implementation
+Whenever the scheduler (i.e. the DRA scheduler plugin) tries to schedule a pod that
+references a `ResourceClaim` with an empty `status.ReservedFor` list, it knows that this
+is the first pod that will be consuming the claim.
+
+If the `spec.ReservedFor` field in the ResourceClaim is not set, the scheduler will handle
+the `ResourceClaim` in the same was as now, and will add the `Pod` to the `ReservedFor` list
+if devices could be allocated for the claim. Any additional pods that reference the `ResourceClaim`
+will also be added to the list.
+
+If the `spec.ReservedFor` field is set, the scheduler will copy this reference to the
+`ReservedFor` list, rather than adding a reference to the `Pod`. It will also update the
+`status.allocation.ReservedForAnyPod` field to `true`. When any other pods referencing
+the `ResourceClaim` is scheduled and the scheduler sees a non-Pod reference in the `ReservedFor`
+list, it will not add a reference to the pod.
+
+##### Deallocation
+The resourceclaim controller will remove Pod references from the `ReservedFor` list just
+like it does now, but it will never remove references to non-Pod resources. Instead, it
+will be the responsibility of the controller/user that created the `ResourceClaim` to
+remove the reference to the non-Pod resource from the `ReservedFor` list when no pods
+are consuming the `ResourceClaim` and no new pods will be created that references
+the `ResourceClaim`.
+
+The resourceclaim controller will then discover that the `ReservedFor` list is empty
+and therefore know that it is safe to deallocate the `ResourceClaim`.
+
+This requires that the controller/user has permissions to update the status
+subresource of the `ResourceClaim`. The resourceclaim controller will also try to detect if
```

Comment:
So, if the other controllers need to do it, then they each need to be granted update permissions to the status. Whereas for the resource claim controller to do it, it needs get/list/watch for that resource type. I think that the latter option has way less opportunity for orphaned resource claims. I would suggest we lead with that as the primary mechanism.

If the resource claim controller sees something in there and doesn't have the right permissions, it can complain in events and logs.

Realistically, this probably means giving the resourceclaim controller the ability to watch deployments, jobs, and statefulsets. And maybe a few others.

Reply:
So I've updated the design here a bit.

What I think might be a possible challenge with making the resource claim controller responsible for removing the reference in the `ReservedFor` list is that we probably can only do this safely when the resource has been deleted (and even then I guess there can be orphaned pods). If a Deployment is scaled down to zero, there will not be any pods using the ResourceClaim, so it could potentially be deallocated. But I don't think the resource claim controller can deallocate just based on the number of replicas, since there will be a race between the deallocation and the possible creation of new pods. If the workload controller is responsible for handling this, I think it should be able to handle this safely.

Reply:
Based on other feedback and thinking more about this, I have changed this to actually put the responsibility on the controllers. Two primary reasons for this:
* It lets us implement a single solution that is available to both in-tree and out-of-tree controllers.
* Controllers have more context about the workloads, so they are in a better position than the resourceclaim controller to decide when deallocation is safe.